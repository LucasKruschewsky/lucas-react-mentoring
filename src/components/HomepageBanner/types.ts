import { IMoviesListData } from '@/data/MockedDataTypes';

export interface IHomeBannerStateToProps {
  selectedMovie?: IMoviesListData;
}

export interface IHomeBannerProps extends IHomeBannerStateToProps {}
