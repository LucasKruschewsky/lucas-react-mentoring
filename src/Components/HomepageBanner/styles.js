import styled from 'styled-components';
import BannerImage from '../../Images/HomepageHero.png';

const BannerContainer = styled.div`
  height: 396px;
  background-image: url(${BannerImage});
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;

  & > div {
    padding: 6.5rem 8rem;
    display: flex;
    flex-direction: column;
  }
`;

const SearchTitle = styled.h1`
  font-size: var(--title-1);
  color: rgb(var(--white));
  margin-bottom: 3rem;
`;

const SearchInputAndButton = styled.div`
  display: flex;
  & > input {
    line-height: 100%;
    padding: 0 1rem;
    flex-grow: 4;
    margin-right: 1rem;
    font-size: var(--body-1);
    color: rgb(var(--white));

    border-color: transparent;
    border-radius: 4px;
    background-color: rgba(var(--secondary), 0.7);
    mix-blend-mode: normal;
  }

  & > button {
    flex-grow: 1;
  }
`;

export { BannerContainer, SearchTitle, SearchInputAndButton };
