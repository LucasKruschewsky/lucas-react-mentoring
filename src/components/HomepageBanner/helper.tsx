import * as React from 'react';
import AppButton from 'Global/styled/AppButton';
import { parseDate } from '@/functions/parseDate';
import { minutesToHours } from '@/functions/minutesToHours';
import { Field, Form, Formik } from 'formik';
import {
  createSearchParams,
  useNavigate,
  useSearchParams,
} from 'react-router-dom';
import { axiosRequest } from '@/functions/axiosRequest';
import { TMovieObject } from '@/store/modules/movieList/types';
import { retrieveAllSearchParams } from '@/functions/retrieveSearchParams';
import {
  SearchBannerContainer,
  SearchTitle,
  SearchInputAndButton,
  SelectedMovieContainer,
  MovieTitleAndRating,
  MovieGenre,
  MovieYearAndDuration,
  MovieDescription,
  MovieImageContainer,
  MovieDetailsContainer,
} from './styles';
import { IHomeBannerProps, ISearchBannerProps } from './types';

export const SearchBanner: React.FunctionComponent<ISearchBannerProps> = ({
  searchQuery,
}) => {
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();

  const handleSubmitSearch = React.useCallback(
    ({ searchField }) => {
      navigate({
        pathname: `/search/${searchField}`,
        search: `?${createSearchParams({
          ...retrieveAllSearchParams(searchParams),
        })}`,
      });
    },
    [navigate, searchParams]
  );

  const initialSearchValue: { searchField: string } = {
    searchField: searchQuery ?? '',
  };

  return (
    <SearchBannerContainer>
      <div>
        <SearchTitle>Find Your Movies</SearchTitle>
        <Formik
          onSubmit={handleSubmitSearch}
          initialValues={initialSearchValue}
        >
          <Form>
            <SearchInputAndButton>
              <Field
                name="searchField"
                id="searchMovie"
                type="text"
                placeholder="What do you want to watch?"
              />
              <AppButton type="submit"> Search </AppButton>
            </SearchInputAndButton>
          </Form>
        </Formik>
      </div>
    </SearchBannerContainer>
  );
};

export const SelectedMovieBanner: React.FunctionComponent<IHomeBannerProps> = ({
  selectedMovieId,
}) => {
  const [selectedMovie, setSelectedMovie] = React.useState<TMovieObject>(null);

  React.useEffect(() => {
    const fetchSelectedMovie = async (): Promise<void> => {
      const response = await axiosRequest(`/movies/${selectedMovieId}`, 'get');
      setSelectedMovie(response.data);
    };

    fetchSelectedMovie();
  }, [selectedMovieId]);

  const parsedRuntime = React.useMemo(
    () => minutesToHours(selectedMovie?.runtime),
    [selectedMovie]
  );

  const parsedDate = React.useMemo(
    () => parseDate(selectedMovie?.release_date),
    [selectedMovie]
  );

  return (
    <SelectedMovieContainer>
      <MovieImageContainer>
        <img
          src={selectedMovie?.poster_path}
          alt={`${selectedMovie?.title} movie`}
        />
      </MovieImageContainer>
      <MovieDetailsContainer id="selected-movie-details-banner">
        <MovieTitleAndRating>
          <h1>{selectedMovie?.title}</h1>
          <div>{selectedMovie?.vote_average}</div>
        </MovieTitleAndRating>
        <MovieGenre>
          {selectedMovie?.genres.map((genre) => (
            <p key={genre}>{genre}</p>
          ))}
        </MovieGenre>
        <MovieYearAndDuration>
          <p>{parsedDate}</p>
          <p>{parsedRuntime}</p>
        </MovieYearAndDuration>
        <MovieDescription>{selectedMovie?.overview}</MovieDescription>
      </MovieDetailsContainer>
    </SelectedMovieContainer>
  );
};
