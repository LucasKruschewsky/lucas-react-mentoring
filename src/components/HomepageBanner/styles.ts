import styled from 'styled-components';
import * as BannerImage from 'Images/HomepageHero.png';
import AppContainer from 'Global/styled/AppContainer';

// SEARCH BANNER STYLES -----------------------------------------

export const SearchBannerContainer = styled(AppContainer)`
  height: 396px;
  background-image: url(${BannerImage});
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;

  & > div {
    padding-top: 6.5rem;
    padding-bottom: 6.5rem;
    display: flex;
    flex-direction: column;
  }

  @media only screen and (min-width: 600px) {
    padding: 0;
    & > div {
      padding: 6.5rem 8rem;
    }
  } ;
`;

export const SearchTitle = styled.h1`
  font-size: var(--title-1);
  color: rgb(var(--white));
  margin-bottom: 3rem;
  text-align: center;

  @media only screen and (min-width: 600px) {
    text-align: left;
  } ;
`;

export const SearchInputAndButton = styled.div`
  display: flex;
  flex-direction: column;

  & > input {
    line-height: 34px;
    padding: 0 1rem;
    flex-grow: 4;
    margin-bottom: 1rem;
    font-size: var(--body-1);
    color: rgb(var(--white));

    border-color: transparent;
    border-radius: 4px;
    background-color: rgba(var(--secondary), 0.7);
    mix-blend-mode: normal;
  }

  & > button {
    flex-grow: 1;
  }

  @media only screen and (min-width: 600px) {
    flex-direction: row;

    & > input {
      margin-right: 1rem;
      margin-bottom: 0;
      line-height: 100%;
    }
  } ;
`;

// SELECTED MOVIE STYLES -----------------------------------------

export const SelectedMovieContainer = styled(AppContainer)`
  display: grid;
  padding-top: 4.5rem;
  grid-auto-flow: column;

  grid-template-columns: repeat(4, minmax(0, 1fr));

  @media only screen and (min-width: 768px) {
    grid-template-columns: repeat(12, minmax(0, 1fr));
  }
`;

export const MovieImageContainer = styled.div`
  grid-column: 2 / span 2;
  order: 2;

  image,
  img,
  svg {
    max-width: 100%;
    max-height: 60vh;
    padding-top: 1rem;
  }

  @media only screen and (min-width: 768px) {
    grid-column: 1 / span 4;
    margin-right: 2rem;
    order: 1;

    image,
    img,
    svg {
      padding-top: 0;
    }
  }
`;

export const MovieDetailsContainer = styled.div`
  grid-column: span 8/ -1;
`;

export const MovieTitleAndRating = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  text-transform: uppercase;
  color: rgb(var(--white));
  h1 {
    font-size: var(--title-1);
    margin-right: 1rem;
  }

  div {
    border: 1px solid white;
    font-size: var(--body-1);
    border-radius: 50%;
    display: flex;

    min-width: 60px;
    min-height: 60px;

    justify-content: center;
    align-items: center;
  }
`;

export const MovieGenre = styled.div`
  margin: 0 0 1rem;
  font-size: 14px;
  color: rgba(var(--white), 0.5);
  display: flex;
  flex-wrap: wrap;

  p {
    margin-right: 1rem;
    margin-top: 0.5rem;
  }
`;

export const MovieYearAndDuration = styled.div`
  display: flex;
  color: rgb(var(--primary));
  font-size: 24px;
  margin: 0 0 1rem;

  & > p {
    margin-right: 3rem;
  }
`;

export const MovieDescription = styled.div`
  color: rgba(var(--white), 0.5);
`;
