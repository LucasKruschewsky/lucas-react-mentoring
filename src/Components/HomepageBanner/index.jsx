import React from 'react';
import AppButton from 'Styles/AppButton';
import { BannerContainer, SearchTitle, SearchInputAndButton } from './styles';

export default function HomepageBanner() {
  return (
    <BannerContainer>
      <div>
        <SearchTitle htmlFor="searchMovie">Find Your Movie</SearchTitle>
        <SearchInputAndButton>
          <input
            name="searchMovie"
            id="searchMovie"
            type="text"
            placeholder="What do you want to watch?"
          />
          <AppButton> Search </AppButton>
        </SearchInputAndButton>
      </div>
    </BannerContainer>
  );
}
