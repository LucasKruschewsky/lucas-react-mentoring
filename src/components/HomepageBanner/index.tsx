import * as React from 'react';
import { useParams, useSearchParams } from 'react-router-dom';
import { SearchBanner, SelectedMovieBanner } from './helper';
import { IHomeBannerProps } from './types';

const HomepageBanner: React.FunctionComponent<IHomeBannerProps> = () => {
  const { searchQuery } = useParams();
  const [searchParams] = useSearchParams();

  return searchParams.has('movie') ? (
    <SelectedMovieBanner selectedMovieId={searchParams.get('movie')} />
  ) : (
    <SearchBanner searchQuery={searchQuery} />
  );
};

export default HomepageBanner;
