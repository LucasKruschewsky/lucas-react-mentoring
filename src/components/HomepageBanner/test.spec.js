import * as React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

// Components
import HomepageBanner from '.';

test('Snapshot test', () => {
  const tree = renderer.create(
    <BrowserRouter>
      <HomepageBanner />
    </BrowserRouter>
  ).toJSON;
  expect(tree).toMatchSnapshot();
});
