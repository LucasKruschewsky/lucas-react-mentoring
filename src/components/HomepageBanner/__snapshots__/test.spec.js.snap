// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `[Function]`;
